### PR TITLE
Continued cleanup and refactoring (breaking changes)

### PR DESCRIPTION
**Pull request by halfak**
* Adds two types of errors: LoginError and APIError
* Removes is_authenticated and related methods from Session
* Adds general file upload capability to post() and removes upload()
* Adds a demo_queries.py script to demonstrate that the library supports login, querying and continuation as well as throws errors correctly when interacting with English Wikipedia's API.
* Adds generalized support for "|" separated param values.  See https://github.com/yuvipanda/python-mwapi/issues/8
* Increments version from 0.1.1 to 0.2.0 to reflect the backwards incompatible changes

### CODE REVIEW

File: mwapi/errors.py [hunk 1]
```diff
@@ -12,10 +12,26 @@ class APIError(RuntimeError):
...
+    def __init__(self, code, info, content):
+        self.code = code
+        self.info = info
+        self.content = content
 
-class LoginError(APIError):
+        super().__init__("{0}: {1} -- {2}".format(code, info, content))
+
+    @classmethod
+    def from_doc(cls, doc):
```
Comment (yuvipanda) [inline mwapi/errors.py]: doc as in documentation?!

File: mwapi/session.py [hunk 2]
```diff
@@ -39,15 +41,21 @@ def __init__(self, host, api_path="/w/api.php", session=None):
...
-        `bool` : Is the session sending authenticated requests
-        """
 
     def _request(self, method, params=None, data=None, files=None):
         resp = self.session.request(method, self.api_url, params=params,
                                     data=data, files=files, stream=True)
-        return resp.json()
+
+        try:
+            doc = resp.json()
```
Comment (yuvipanda) [inline mwapi/session.py]: doc feels like a confusing variable name here - I usually use a variant of 'result', or spell out document fully.

File: mwapi/session.py [hunk 3]
```diff
@@ -84,66 +92,42 @@ def logout(self):
         Logs out of the session with MediaWiki
         """
         self.post(action='logout')
-        self.is_authenticated = False
```
Comment (yuvipanda) [inline mwapi/session.py]: So this is basically stripping away password auth, and will add support for mwoauth in some way in the future. Is that correct?

**Comment by yuvipanda** [comment]
General +1, but might need to rebase to trigger Travis?

File: mwapi/session.py (hunk 2 above)
Comment (halfak) [inline mwapi/session.py]: I always thought that "doc" as in json document was intuitive.  "result" could be all sorts of different types of things.  I'm down with spelling out "document" completely, but that seems verbose when "doc" was assumed to stand for "document" in the first place.    See "token_doc" and "login_doc" below.

File: mwapi/session.py (hunk 3 above)
Comment (halfak) [inline mwapi/session.py]: password auth is still here.  I think that is important.  The boolean local variable is problematic for a lot of reasons.  Essentially you don't know when it is out of date.   But yeah, I figured that a a persistent auth-cookie should be dropped and OAuth support used instead.

File: mwapi/errors.py (hunk 1 above)
Comment (halfak) [inline mwapi/errors.py]: doc as in document.

File: mwapi/session.py (hunk 2 above)
Comment (yuvipanda) [inline mwapi/session.py]: Fair enough, I guess as long as it's consistently 'doc' everywhere...

**Comment by halfak** [comment]
@yuvipanda remaining flake8 issues are not actually issues.  Want me to do anything else.

**Comment by yuvipanda** [comment]
Was hoping to convince you to do
https://github.com/wikimedia/operations-software-tools-webservice/blob/master/toollabs/webservice/__init__.py
which is more explicit and takes away the warning as well. The alternative
is to comment # noqa at the end of the imports which ignores those imports,
but I like the more explicit __all__ version better.

In general it's a good idea to have the minor tests we have passing, I
guess.

On Thu, Aug 27, 2015 at 4:21 PM, Aaron Halfaker <notifications@github.com>
wrote:

> @yuvipanda <https://github.com/yuvipanda> remaining flake8 issues are not
> actually issues. Want me to do anything else.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/yuvipanda/python-mwapi/pull/11#issuecomment-135450586>
> .
>



-- 
Yuvi Panda T
http://yuvi.in/blog

**Comment by halfak** [comment]
@yuvipanda {{done}}

**Comment by yuvipanda** [comment]
Woo!

**Pull request merged**